mypy/typestate.py

```diff
@@ -62,3 +91,99 @@ def record_subtype_cache_entry(cls, left: Instance, right: Instance) -> None:
     @classmethod
     def record_proper_subtype_cache_entry(cls, left: Instance, right: Instance) -> None:
         cls._subtype_caches_proper.setdefault(right.type, set()).add((left, right))
+
+    @classmethod
+    def reset_protocol_deps(cls) -> None:
+        cls.proto_deps = {}
+        cls._attempted_protocols = {}
+        cls._checked_against_members = {}
+        cls._rechecked_types = set()
+
+    @classmethod
+    def record_protocol_subtype_check(cls, left_type: TypeInfo, right_type: TypeInfo) -> None:
+        assert right_type.is_protocol
+        cls._rechecked_types.add(left_type)
+        cls._attempted_protocols.setdefault(left_type, set()).add(right_type.fullname())
+        cls._checked_against_members.setdefault(left_type,
+                                                set()).update(right_type.protocol_members)
+
+    @classmethod
+    def _snapshot_protocol_deps(cls) -> Dict[str, Set[str]]:
+        """Collect protocol attribute dependencies found so far from registered subtype checks.
+
+        There are three kinds of protocol dependencies. For example, after a subtype check:
+
+            x: Proto = C()
+
+        the following dependencies will be generated:
+            1. ..., <SuperProto[wildcard]>, <Proto[wildcard]> -> <Proto>
+            2. ..., <B.attr>, <C.attr> -> <C> [for every attr in Proto members]
+            3. <C> -> Proto  # this one to invalidate the subtype cache
+
+        The first kind is generated immediately per-module in deps.py. While two other kinds are
+        generated here after all modules are type checked anf we have recorded all the subtype
+        checks.
```

Comment:
Can you give examples where each of these dependency kinds are needed to produce correct type checking results?